rulewerk-vlog/src/main/java/org/semanticweb/rulewerk/reasoner/vlog/VLogFastQueryResultIterator.java

```diff
@@ -0,0 +1,234 @@
+	 */
+	private final VLog vLog;
+	/**
+	 * VLog ids of the previous tuple, with the last id omitted (since it is not
+	 * useful in caching).
+	 */
+	private long[] prevIds = null;
+	/**
+	 * RuleWerk terms corresponding to the previously fetched tuple, with the last
+	 * term omitted.
+	 */
+	private Term[] prevTerms = null;
+	/**
+	 * True if this is the first result that is returned.
+	 */
+	boolean firstResult = true;
+	/**
+	 * Size of the tuples returned in this result.
+	 */
+	int resultSize = -1;
+	/**
+	 * Cache mapping ids to terms.
+	 */
+	// final SimpleLruMap termCache;
+	final OrderedTermCache termCache;
+
+	private final Correctness correctness;
+
+	/**
+	 * Create a new {@link VLogFastQueryResultIterator}.
+	 * 
+	 * @param queryResultIterator
+	 * @param materialisationState
+	 * @param vLog
+	 */
+	public VLogFastQueryResultIterator(final karmaresearch.vlog.QueryResultIterator queryResultIterator,
+			final Correctness materialisationState, final VLog vLog) {
+		this.vLogQueryResultIterator = queryResultIterator;
+		this.correctness = materialisationState;
+		this.vLog = vLog;
+		// this.termCache = new SimpleLruMap(256, 64000);
+		this.termCache = new OrderedTermCache(130000);
+	}
+
+	@Override
+	public boolean hasNext() {
+		return this.vLogQueryResultIterator.hasNext();
+	}
+
+	@Override
+	public QueryResult next() {
+		final Term[] terms;
+		long[] idTuple = vLogQueryResultIterator.next();
+		terms = new Term[idTuple.length];
+
+		if (firstResult) {
+			resultSize = terms.length;
+			prevIds = new long[resultSize - 1];
+			prevTerms = new Term[resultSize - 1];
+		}
+
+		int i = 0;
+		for (long id : idTuple) {
+			if (!firstResult && i < resultSize - 1 && prevIds[i] == id) {
+				terms[i] = prevTerms[i];
+			} else if (resultSize == 1) { // caching pointless for unary queries
+				terms[i] = computeTerm(id);
+			} else {
+				Term term = this.termCache.get(id);
+				if (term == null) {
+					term = computeTerm(id);
+					this.termCache.put(id, term);
+				}
+				terms[i] = term;
+				if (i < resultSize - 1) {
+					prevTerms[i] = term;
+					prevIds[i] = id;
+				}
+			}
+			i++;
+		}
+
+		firstResult = false;
+		return new QueryResultImpl(Arrays.asList(terms));
+	}
+
+	/**
+	 * Compute the {@link Term} for a given VLog id.
+	 * 
+	 * @param id
+	 * @return
+	 */
+	Term computeTerm(long id) {
+		try {
+			String s = vLog.getConstant(id);
+			// This internal handling is copied from VLog's code in {@link
+			// karmaresearch.vlog.TermQueryResultIterator}.
+			// TODO: the string operation to make null names should possibly be provided by
+			// VLog rather than being hardcoded here?
+			if (s == null) {
+				return new NamedNullImpl("" + (id >> 40) + "_" + ((id >> 32) & 0377) + "_" + (id & 0xffffffffL));
+			} else {
+				return VLogToModelConverter.toConstant(s);
+			}
+		} catch (NotStartedException e) {
+			// Should not happen, we just did a query ...
+			throw new RuntimeException(e);
```

Comment:
throw `RulewerkRuntimeException` instead.

Reply:
Why do we actually have RulewerkRuntimeException? Note that this line will never be executed in practice either.

Reply:
Actually, it should rather throw `ReasonerStateException` (which `RulewerkRuntimeException` is a base class of); this is inconsistently used throughout the code base, however.